src/server/accounting.c

```diff
@@ -353,244 +390,138 @@ get_walltime(job *jp, int res)
+		return;
 #endif
-	i = strlen(pb);
-	pb  += i;
-	len -= i;
 
 	/* group */
-	nd = sizeof(GROUP_FMT) + strlen(pjob->ji_wattr[(int)JOB_ATR_egroup].at_val.at_str);
-	if (nd > len)
-		if (grow_acct_buf(&pb, &len, nd) == -1)
-			return (pb);
+	if (!pbs_strcat(buf, len, GROUP_FMT)) 
+		return;
 #ifdef WIN32
-	(void)snprintf(pb, len, GROUP_FMT,
-		replace_space(pjob->ji_wattr[(int)JOB_ATR_egroup].at_val.at_str,
-		acctlog_spacechar));
+	win_str = replace_space(pjob->ji_wattr[(int)JOB_ATR_egroup].at_val.at_str, acctlog_spacechar);
+	if (!pbs_strcat(buf, len, win_str))
+		return;
 #else
-	(void)snprintf(pb, len, GROUP_FMT,
-		pjob->ji_wattr[(int)JOB_ATR_egroup].at_val.at_str);
+	if (!pbs_strcat(buf, len, pjob->ji_wattr[(int)JOB_ATR_egroup].at_val.at_str))
+		return;
 #endif
-	i = strlen(pb);
-	pb  += i;
-	len -= i;
 
 	/* account */
 	if (pjob->ji_wattr[(int)JOB_ATR_account].at_flags & ATR_VFLAG_SET) {
-		nd = sizeof(ACCOUNT_FMT) + strlen(pjob->ji_wattr[(int)JOB_ATR_account].at_val.at_str);
-		if (nd > len)
-			if (grow_acct_buf(&pb, &len, nd) == -1)
-				return (pb);
 #ifdef WIN32
-		(void)snprintf(pb, len, ACCOUNT_FMT,
-			replace_space(
-			pjob->ji_wattr[(int)JOB_ATR_account].at_val.at_str,
-			acctlog_spacechar));
+		win_str = replace_space(pjob->ji_wattr[(int)JOB_ATR_account].at_val.at_str, acctlog_spacechar);
+		if (cpy_quote_value(buf, len, win_str, ACCOUNT_FMT) == -1)
+			return;
 #else
-		(void)snprintf(pb, len, ACCOUNT_FMT,
-			pjob->ji_wattr[(int)JOB_ATR_account].at_val.at_str);
+		if (cpy_quote_value(buf, len, pjob->ji_wattr[(int)JOB_ATR_account].at_val.at_str, ACCOUNT_FMT) == -1)
+			return;
 #endif
-		i = strlen(pb);
-		pb  += i;
-		len -= i;
 	}
 
 	/* project */
 	if (pjob->ji_wattr[(int)JOB_ATR_project].at_flags & ATR_VFLAG_SET) {
-		char *projstr;
-
-		projstr =  pjob->ji_wattr[(int)JOB_ATR_project].at_val.at_str;
-		/* using PROJECT_FMT1 if projstr needs to be quoted; otherwise, PROJECT_FMT2 */
-		nd = sizeof(PROJECT_FMT1) + strlen(projstr);
-		if (nd > len)
-			if (grow_acct_buf(&pb, &len, nd) == -1)
-				return (pb);
-		if (strchr(projstr, ' ') != NULL) {
-			(void)snprintf(pb, len, PROJECT_FMT1, projstr);
-		} else {
-			(void)snprintf(pb, len, PROJECT_FMT2, projstr);
-		}
-		i = strlen(pb);
-		pb  += i;
-		len -= i;
+		if (cpy_quote_value(buf, len, pjob->ji_wattr[(int)JOB_ATR_project].at_val.at_str, PROJECT_FMT) == -1)
+			return;
 	}
 
 	/* accounting_id */
 	if (pjob->ji_wattr[(int)JOB_ATR_acct_id].at_flags & ATR_VFLAG_SET) {
-		nd = sizeof(ACCOUNTING_ID_FMT) + strlen(pjob->ji_wattr[(int)JOB_ATR_acct_id].at_val.at_str);
-		if (nd > len)
-			if (grow_acct_buf(&pb, &len, nd) == -1)
-				return (pb);
-		(void)snprintf(pb, len, ACCOUNTING_ID_FMT,
-			pjob->ji_wattr[(int)JOB_ATR_acct_id].at_val.at_str);
-		i = strlen(pb);
-		pb  += i;
-		len -= i;
+		if (!pbs_strcat(buf, len, ACCOUNTING_ID_FMT)) 
+			return;
+		if (!pbs_strcat(buf, len, pjob->ji_wattr[(int)JOB_ATR_acct_id].at_val.at_str))
```

Comment:
Some of the attributes will be parsed via PBS itself where there can not be any possibilities that it can be malicious or can be altered by someone else in terms of security so copy_quote_value() has been only used where the values can be altered.